store/commitment/iavl/tree.go

```diff
@@ -83,6 +93,140 @@ func (t *IavlTree) Prune(version uint64) error {
+		return fmt.Errorf("failed to write store name: %w", err)
+	}
+
+	for {
+		node, err := exporter.Next()
+		if err == iavl.ErrorExportDone {
+			break
+		} else if err != nil {
+			return fmt.Errorf("failed to get the next export node: %w", err)
+		}
+		if err = protoWriter.WriteMsg(&snapshottypes.SnapshotItem{
+			Item: &snapshottypes.SnapshotItem_IAVL{
+				IAVL: &snapshottypes.SnapshotIAVLItem{
+					Key:     node.Key,
+					Value:   node.Value,
+					Height:  int32(node.Height),
+					Version: node.Version,
+				},
+			},
+		}); err != nil {
+			return fmt.Errorf("failed to write iavl node: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Restore implements snapshottypes.CommitSnapshotter.
+// returns next snapshot item and error.
+func (t *IavlTree) Restore(version uint64, format uint32, protoReader protoio.Reader, chStorage chan<- *store.KVPair) (snapshottypes.SnapshotItem, error) {
+	var importer *iavl.Importer
+	var snapshotItem snapshottypes.SnapshotItem
+loop:
+	for {
+		snapshotItem = snapshottypes.SnapshotItem{}
+		err := protoReader.ReadMsg(&snapshotItem)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return snapshottypes.SnapshotItem{}, fmt.Errorf("invalid protobuf message: %w", err)
+		}
+
+		switch item := snapshotItem.Item.(type) {
+		case *snapshottypes.SnapshotItem_Store:
+			t.storeKey = item.Store.Name
+			importer, err = t.tree.Import(int64(version))
+			if err != nil {
+				return snapshottypes.SnapshotItem{}, fmt.Errorf("failed to import tree for version %d: %w", version, err)
+			}
+			defer importer.Close()
+
+		case *snapshottypes.SnapshotItem_IAVL:
+			if importer == nil {
+				return snapshottypes.SnapshotItem{}, fmt.Errorf("received IAVL node item before store item")
+			}
+			if item.IAVL.Height > math.MaxInt8 {
+				return snapshottypes.SnapshotItem{}, fmt.Errorf("node height %v cannot exceed %v",
+					item.IAVL.Height, math.MaxInt8)
+			}
+			node := &iavl.ExportNode{
+				Key:     item.IAVL.Key,
+				Value:   item.IAVL.Value,
+				Height:  int8(item.IAVL.Height),
+				Version: item.IAVL.Version,
+			}
+			// Protobuf does not differentiate between []byte{} and nil, but fortunately IAVL does
+			// not allow nil keys nor nil values for leaf nodes, so we can always set them to empty.
+			if node.Key == nil {
+				node.Key = []byte{}
+			}
+			if node.Height == 0 {
+				if node.Value == nil {
+					node.Value = []byte{}
+				}
+				// If the node is a leaf node, it will be written to the storage.
+				chStorage <- &store.KVPair{
+					StoreKey: t.storeKey,
+					Key:      node.Key,
+					Value:    node.Value,
+				}
+			}
+			err := importer.Add(node)
+			if err != nil {
+				return snapshottypes.SnapshotItem{}, fmt.Errorf("failed to add node to importer: %w", err)
+			}
+		default:
+			break loop
+		}
+	}
+
+	if importer != nil {
+		err := importer.Commit()
+		if err != nil {
+			return snapshottypes.SnapshotItem{}, fmt.Errorf("failed to commit importer: %w", err)
+		}
+		importer.Close()
```

Comment:
Should this be deferred before the call to `Commit`?

Reply:
removed it, since it is deferred when creating